### ISSUE TITLE
Constructor inconsistency

### ISSUE THREAD

**Issue by dgrigonis** [pydata/xarray#7901]
### What happened?

Ambiguities in xr.DataArray construction

### What did you expect to happen?

I expect that printing and resolving coords is based on `dims`

If not supplied, default dims is a list of [`dim_0`, `dim_1`]

Then, coords (if dictionary) is mapped either to default or supplied `dims`

As you can see from my example, it does seem inconsistent in both construction and printing.

Another point is that order in `coords` dictionary matters (as can be seen from example below). Although, currently in python dict is ordered by default, this does prevent `easy` and `neat` construction, while running risk inconsistencies and bugs, when supplying coords from json or other unordered mapping sources.

### Minimal Complete Verifiable Example

```Python
arr = np.ones((2, 3))
# THESE 2 WORK
xarr = xr.DataArray(arr, coords=dict(dim_0=np.arange(2), dim_1=list('abc')))
xarr = xr.DataArray(arr, coords=dict(dim_1=list('abc')))

# THIS WORKS TOO
xarr = xr.DataArray(arr, coords=dict(dim_1=list('abc'), dim_0=np.arange(2)), dims=['dim_0', 'dim_1'])
# But print order of dims is based on coords, not on dims `print(xarr)`

# THIS DOESN'T
xarr = xr.DataArray(arr, coords=dict(dim_1=list('abc'), dim_0=np.arange(2)))


# BEHAVIOR I WOULD EXPECT
xarr = xr.DataArray(arr, coords=dict(dim_1=list('abc'), dim_0=np.arange(2)))
# Is equivalent to:
xarr = xr.DataArray(arr, coords=dict(dim_0=np.arange(2), dim_1=list('abc')))
```

### MVCE confirmation

- [ ] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [ ] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [ ] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Environment

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.11.3 (main, Apr  7 2023, 19:30:05) [Clang 13.0.0 (clang-1300.0.29.30)]
python-bits: 64
OS: Darwin
OS-release: 20.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: ('en_GB', 'UTF-8')
libhdf5: None
libnetcdf: None

xarray: 2023.5.0
pandas: 2.0.1
numpy: 1.24.2
scipy: 1.10.1
netCDF4: None
pydap: None
h5netcdf: None
h5py: None
Nio: None
zarr: None
cftime: None
nc_time_axis: None
PseudoNetCDF: None
iris: None
bottleneck: 1.3.7
dask: None
distributed: None
matplotlib: 3.7.1
cartopy: None
seaborn: None
numbagg: 0.2.2
fsspec: 2023.5.0
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 67.6.1
pip: 23.1.2
conda: None
pytest: None
mypy: None
IPython: 8.12.0
sphinx: None
</details>

**Comment by dgrigonis** [comment]
I agree, but what I am arguing with is that this method of coordinate inference should not be allowed:
```
arr = np.ones((2, 2))
arr1 = xr.DataArray(arr, coords=dict(dim_0=np.arange(2), dim_1=list("ab")))
arr2 = xr.DataArray(arr, coords=dict(dim_1=list("ab"), dim_0=np.arange(2)))
```
It relies on order of the dictionary, which is error prone due to possibility of it being ported from say json, where it is unsorted.

What I am proposing is that if `dims` is not given, ['dim_0', 'dim_1'] etc is assumed regardless if `len(coords)` matches number of dimensions or not.

See my PR: https://github.com/pydata/xarray/pull/7903

**Comment by max-sixty** [comment]
xarray doesn't generally rely on the _order_ of dimensions. So I don't think this is a significant issue.

Are there common cases having an unexpected order of dimensions would be an issue?

**Comment by dgrigonis** [comment]
One common case that I have encountered is when array is serialised to json (or other protocol, which does not preserve order mapping objects) and then serialising it back.

Of course, knowing this, there are measures to prevent this from happening. But my point is that preventing such behaviour would eliminate the risk of this happening.

In my example below, the order is "accidentally" preserved, but that might not always be the case. I am not going to test all serialisers out there to find which ones fail, but the fact is that "JSON Dictionary doesn't have any order.". And I think my proposal is a relatively small cost to eliminate this risk, given this concerns all serialisers with such property.

```python
import json
import xarray as xr

arr = np.ones((2, 2))
xarr = xr.DataArray(arr, coords=dict(dim_0=np.arange(2), dim_1=list("ab")))

obj = dict(values=xarr.values.tolist(), coords={k: v.values.tolist() for k, v in xarr.coords.items()})
serialised = json.loads(json.dumps(obj))    # If this part was to mess the order up.

xarr_des = xr.DataArray(deserialised['values'], coords=deserialised['coords'])
```

**Comment by headtr1ck** [comment]
Why don't you serialize the dims as well?
Afaik json Arrays are ordered.

Not sure if we need to change the constructor for such a corner case. Especially since this would be quite a breaking change, I often rely on this ordering for creating small test dataarrays.

**Comment by dgrigonis** [comment]
Yes, as I said, there are more than one way to prevent this. I proposed this with intent of making it more robust, but if the cost is too great, that is understandable. If that is so, I will close the issue and cancel PR.

**Comment by max-sixty** [comment]
Right — the dims are named, so we don't need to rely on the ordering — that's one of the advantages of named dims!

The example above discards the names, so hence the inconsistency.

I do appreciate the push for consistency, and I recognize the constructor's behavior has some variance from whether the number of dims matches the number of coords.

But on balance that presents quite a small issue relative to the large change we'd need to make here...
